big_tests/tests/ct_helper.erl

```diff
@@ -25,24 +44,48 @@ repeat_all_until_all_ok(GroupDefs) ->
 %% However, if `repeat_until_all_ok` is already defined in Properties it will not be redefined.
 %% This allows you to declaratively override Retries for a particular group definition,
 %% while allowing repeat_all_until_all_ok/2 to add the default for the remaining groups.
+-spec repeat_all_until_all_ok([group_def() | group_def_dirty() | group_def_incomplete()],
+                              repeat_num()) ->
+    [group_def()].
 repeat_all_until_all_ok(GroupDefs, Retries) ->
-    [ {Name, repeat_type(repeat_until_all_ok, Retries, Properties), Tests}
+    [ {Name, maybe_add_repeat_type(repeat_until_all_ok, Retries, Properties), Tests}
       || {Name, Properties, Tests} <- prepare_group_defs(GroupDefs) ].
 
-repeat_type(RepeatType, Retries, Properties) ->
-    %% lists:keyfind expects tuples, but Properties might contain just RepeatType atoms,
-    %% not their tuple forms. Make sure they're expanded to {RepeatType, true} if they
-    %% are present.
-    case lists:keyfind(RepeatType, 1, proplists:unfold(Properties)) of
-        false ->
-            [{RepeatType, Retries} | Properties];
-        {RepeatType, _} ->
-            ct:pal("~s present in ~p - leaving as is", [RepeatType, Properties]),
-            Properties
+-spec maybe_add_repeat_type(repeat_type(), repeat_num(), group_props()) -> group_props().
+maybe_add_repeat_type(RepeatType, Retries, Properties) ->
+    case lists:any(fun(El) -> proplists:is_defined(El, Properties) end, all_repeat_modes()) of
+        true ->
+            Properties;
+        _ ->
+            [{RepeatType, Retries} | Properties]
     end.
 
-%% @doc Expand 2-element group definitions into 3-element group definitions.
+%% @doc Expand dirty group definitions into 3-elements clean group definitions.
+-spec prepare_group_defs([group_def() | group_def_dirty() | group_def_incomplete()]) ->
+    [group_def()].
 prepare_group_defs(GroupDefs) ->
-    lists:map(fun ({Name, Properties, Tests}) -> {Name, Properties, Tests};
-                  ({Name, Tests}) -> {Name, [], Tests} end,
+    lists:map(fun ({Name, Tests}) -> {Name, [], Tests};
+                  ({Name, Properties, Tests}) ->
+                      {Name, clean_properties(Properties), Tests}
+              end,
               GroupDefs).
+
+%% Properties might contain just RepeatType atoms,
+%% so we expand them to {RepeatType, sensible_maximum} if so.
+% -spec clean_properties(group_props() | group_props_dirty()) -> group_props().
+clean_properties(Props) when is_list(Props) ->
+    Max = sensible_maximum_repeats(),
+    proplists:expand(
+      lists:flatten(
+        [[{R, [{R, Max}]}, {{R, forever}, [{R, Max}]}]
```

Comment:
While it's not a universal convention, I tend to wrap comprehensions in brackets separated by spaces from the content to make them more clearly distinguishable from ordinary lists - especially, when items built in the comprehension are complex. For example:

```erlang
[ [A, B]
  || {A, B} <- ListOfTuples ].
```

Instead of:

```erlang
[[A, B]
 || {A, B} <- ListOfTuples].
```

Maybe it's just me, but it immediately seems clearer.

Reply:
Right, seems fair. Will do next time 🙂